Superbuild: aarch64/Linux misdetected as x64 (UseBoost.cmake probes with Apple-only "-arch" flag)

## Summary

On Linux/aarch64, SCIRun's Boost configuration misdetects the CPU architecture and falls back to a hardcoded `x64` platform suffix. `Superbuild/UseBoost.cmake` probes the architecture with the Apple-Clang-only `-arch` flag, which GCC/Clang on Linux reject, so **both** the x86_64 and arm64 probes fail. The failed arm64 probe drops the build into the `x64` branch, and that suffix is baked into the expected Boost static-library filenames — so an aarch64 host looks for the wrong libraries and the build breaks. The arm64 path is effectively unreachable on Linux; the architecture check has no correct outcome there.

Reproduced against `origin/master` @ 148b3247b (2026-07-24); `Superbuild/UseBoost.cmake` unchanged since 5f1729b (2026-04-29).

## Environment

- **Host:** NVIDIA GB10 (Grace-Blackwell), aarch64 — Cortex-X925/A725, 20 cores
- **OS:** Ubuntu 24.04, kernel 6.17 (aarch64)
- **Compiler:** GCC 13.3.0 (Ubuntu 13.3.0-6ubuntu2~24.04.1)
- **CMake:** 3.28.3

## Offending code

`Superbuild/UseBoost.cmake`, lines 12-39:

```cmake
check_c_compiler_flag("-arch x86_64" x86_64Supported)
check_c_compiler_flag("-arch arm64" arm64Supported)
...
  if(${arm64Supported})
    set(PLATFORM_SUFFIX "a64")
  else()
    set(PLATFORM_SUFFIX "x64")
  endif()
```

`PLATFORM_SUFFIX` is then used to build the expected Boost library filenames (same file, lines 47-55), e.g. `libboost_thread-mt-${PLATFORM_SUFFIX}.a`.

`-arch <arch>` is an Apple Clang driver flag. GCC and upstream Clang on Linux do not accept it, so `check_c_compiler_flag` records failure for **both** probes regardless of the real CPU. The arm64 branch is therefore unreachable on Linux, and aarch64 hosts are silently treated as `x64`.

## Reproduction

Minimal CMake harness running the exact logic from `UseBoost.cmake` on the aarch64 host:

```cmake
cmake_minimum_required(VERSION 3.10)
project(archrepro C CXX)
include(CheckCCompilerFlag)

check_c_compiler_flag("-arch x86_64" x86_64Supported)
check_c_compiler_flag("-arch arm64" arm64Supported)
if(${arm64Supported})
  set(PLATFORM_SUFFIX "a64")
else()
  set(PLATFORM_SUFFIX "x64")
endif()

message(STATUS "CMAKE_SYSTEM_PROCESSOR = ${CMAKE_SYSTEM_PROCESSOR}")
message(STATUS "arm64Supported         = ${arm64Supported}")
message(STATUS "PLATFORM_SUFFIX        = ${PLATFORM_SUFFIX}")
```

Actual output (aarch64 host):

```
-- Performing Test x86_64Supported - Failed
-- Performing Test arm64Supported - Failed
-- CMAKE_SYSTEM_PROCESSOR = aarch64
-- arm64Supported         =
-- PLATFORM_SUFFIX        = x64
```

`CMakeError.log` records the underlying compiler diagnostic for the probe:

```
cc: error: unrecognized command-line option '-arch'; did you mean '-march='?
```

So on an aarch64 machine the build resolves `PLATFORM_SUFFIX = x64` and looks for `libboost_<lib>-mt-x64.a`. The arm64 path (`-a64`) is never reachable on Linux.

## Expected vs. actual

- **Expected:** on aarch64 Linux, the platform is detected as arm64 and the build proceeds with the correct Boost library naming.
- **Actual:** both `-arch` probes fail; the platform is silently labeled `x64`; Boost library filename resolution is wrong for the host.

## Suggested fix

Detect the architecture from `CMAKE_SYSTEM_PROCESSOR` rather than probing an Apple-only compiler flag. This is correct on all three platforms (the macOS universal-build case can still special-case `-arch`). For example:

```cmake
if(APPLE)
  # macOS may build universal/cross; keep the -arch probe here if needed.
  check_c_compiler_flag("-arch arm64" arm64Supported)
else()
  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|arm64")
    set(arm64Supported TRUE)
  else()
    set(arm64Supported FALSE)
  endif()
endif()

if(arm64Supported)
  set(PLATFORM_SUFFIX "a64")
else()
  set(PLATFORM_SUFFIX "x64")
endif()
```

Minor: `if(${arm64Supported})` dereferences an empty variable when the probe is undefined; `if(arm64Supported)` is the safer form.

## Notes

Reporter (Cliff Miller, LAAF / Gutlaf) is happy to test a patch on the GB10 / aarch64 box and confirm the full superbuild completes.

### Reproduced on GitHub-hosted Linux/aarch64 CI

Added a fast, dependency-free regression check ([branch `claude/arm64-boost-suffix-repro`](https://github.com/SCIInstitute/SCIRun/tree/claude/arm64-boost-suffix-repro)) that `include()`s the real `Superbuild/UseBoost.cmake` and asserts `PLATFORM_SUFFIX` matches the host CPU. It runs a matrix over `ubuntu-24.04` (x86_64) and `ubuntu-24.04-arm` (aarch64) — configure-time only, no Boost build.

Result matches the report exactly:

| Job | Runner | Result |
|---|---|---|
| PLATFORM_SUFFIX on x86_64 | `ubuntu-24.04` | ✅ pass (`x64`) |
| PLATFORM_SUFFIX on aarch64 | `ubuntu-24.04-arm` | ❌ fail |

**Failing job log:** https://github.com/SCIInstitute/SCIRun/actions/runs/30243813486/job/89906394098

Key lines from that log (GCC 13.3.0, aarch64):

```
-- Performing Test x86_64Supported - Failed
-- Performing Test arm64Supported - Failed
-- CMAKE_SYSTEM_PROCESSOR = aarch64
-- PLATFORM_SUFFIX        = x64
CMake Error: Boost PLATFORM_SUFFIX mismatch: got 'x64', expected 'a64' for
CMAKE_SYSTEM_PROCESSOR='aarch64'.  UseBoost.cmake misdetected the CPU
```

This is the RED half of a red-green fix; the fix is in progress on a separate branch.

--- a/.github/workflows/boost-platform-suffix.yml
+++ b/.github/workflows/boost-platform-suffix.yml
@@ -1,0 +1,37 @@
+name: Boost platform-suffix detection
+
+# Fast, dependency-free regression check for the Boost architecture detection
+# in Superbuild/UseBoost.cmake. It only runs `cmake` (the assertion fires at
+# configure time), so it finishes in seconds and never builds Boost or SCIRun.
+
+on:
+  push:
+    branches: ['**']
+    paths:
+      - 'Superbuild/UseBoost.cmake'
+      - 'Superbuild/Tests/BoostPlatformSuffix/**'
+      - '.github/workflows/boost-platform-suffix.yml'
+  pull_request:
+    paths:
+      - 'Superbuild/UseBoost.cmake'
+      - 'Superbuild/Tests/BoostPlatformSuffix/**'
+      - '.github/workflows/boost-platform-suffix.yml'
+  workflow_dispatch:
+
+jobs:
+  platform-suffix:
+    name: PLATFORM_SUFFIX on ${{ matrix.arch }}
+    runs-on: ${{ matrix.runner }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - arch: x86_64
+            runner: ubuntu-24.04
+          - arch: aarch64
+            runner: ubuntu-24.04-arm
+    steps:
+      - uses: actions/checkout@v4
+      - name: Configure test (asserts PLATFORM_SUFFIX matches host arch)
+        run: |
+          cmake -S Superbuild/Tests/BoostPlatformSuffix -B build-suffix-check

--- a/.github/workflows/ccpp.yml
+++ b/.github/workflows/ccpp.yml
@@ -48,3 +48,29 @@ jobs:
       runner: ubuntu-latest
       variant: gui
       with-ospray: true
+
+  # arm64 (aarch64) coverage. GitHub-hosted Linux/arm64 runners
+  # (ubuntu-24.04-arm) are free for public repositories. These mirror the x86_64
+  # jobs above so the nightly/master matrix exercises the arm64 build path --
+  # e.g. the Boost PLATFORM_SUFFIX arch detection fixed in this branch (#2583).
+  linux-gui-arm:
+    uses: ./.github/workflows/reusable-build.yml
+    with:
+      runner: ubuntu-24.04-arm
+      qt-version: ''
+      variant: gui
+      build-with-python: true
+
+  linux-headless-arm:
+    uses: ./.github/workflows/reusable-build.yml
+    with:
+      runner: ubuntu-24.04-arm
+      variant: headless
+      build-testing: true
+
+  linux-gui-nonpython-arm:
+    uses: ./.github/workflows/reusable-build.yml
+    with:
+      runner: ubuntu-24.04-arm
+      variant: gui
+      build-with-python: false

--- a/Superbuild/Tests/BoostPlatformSuffix/CMakeLists.txt
+++ b/Superbuild/Tests/BoostPlatformSuffix/CMakeLists.txt
@@ -1,0 +1,41 @@
+# Regression test for the Boost PLATFORM_SUFFIX architecture detection in
+# Superbuild/UseBoost.cmake.
+#
+# UseBoost.cmake historically probed the CPU architecture with the Apple-only
+# "-arch arm64" compiler flag. GCC/Clang on Linux reject that flag, so on
+# Linux/aarch64 the probe failed and the build fell back to the "x64" suffix --
+# baking the wrong architecture into the expected Boost library filenames.
+# See the aarch64/GB10 bug report.
+#
+# This test includes the REAL UseBoost.cmake (SCI_BOOST_LIBRARY is left unset,
+# so its per-library foreach is a no-op) and asserts that PLATFORM_SUFFIX
+# matches the host architecture. It goes red on aarch64 today and green once
+# the detection is fixed; it stays green on x86_64.
+
+cmake_minimum_required(VERSION 3.10)
+project(BoostPlatformSuffixTest C CXX)
+
+include(CheckCCompilerFlag)
+
+# Path to the file under test: <repo>/Superbuild/UseBoost.cmake
+include(${CMAKE_CURRENT_SOURCE_DIR}/../../UseBoost.cmake)
+
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|arm64")
+  set(EXPECTED_SUFFIX "a64")
+else()
+  set(EXPECTED_SUFFIX "x64")
+endif()
+
+message(STATUS "CMAKE_SYSTEM_PROCESSOR = ${CMAKE_SYSTEM_PROCESSOR}")
+message(STATUS "PLATFORM_SUFFIX        = ${PLATFORM_SUFFIX}")
+message(STATUS "EXPECTED_SUFFIX        = ${EXPECTED_SUFFIX}")
+
+if(NOT PLATFORM_SUFFIX STREQUAL EXPECTED_SUFFIX)
+  message(FATAL_ERROR
+    "Boost PLATFORM_SUFFIX mismatch: got '${PLATFORM_SUFFIX}', "
+    "expected '${EXPECTED_SUFFIX}' for CMAKE_SYSTEM_PROCESSOR="
+    "'${CMAKE_SYSTEM_PROCESSOR}'. UseBoost.cmake misdetected the CPU "
+    "architecture.")
+endif()
+
+message(STATUS "PASS: Boost PLATFORM_SUFFIX matches host architecture.")

--- a/Superbuild/UseBoost.cmake
+++ b/Superbuild/UseBoost.cmake
@@ -9,10 +9,20 @@ if(UNIX)
   add_definitions(-DBOOST_NO_CXX11_ALLOCATOR)
 endif()
 
-check_c_compiler_flag("-arch x86_64" x86_64Supported)
-#message("x86_64Supported=${x86_64Supported}")
-check_c_compiler_flag("-arch arm64" arm64Supported)
-#message("arm64Supported=${arm64Supported}")
+# Detect whether we are targeting arm64. On Apple platforms the "-arch" driver
+# flag is valid and also covers universal/cross builds, so probe it there.
+# Elsewhere (GCC/Clang on Linux) "-arch" is rejected and both probes would fail,
+# silently mislabeling aarch64 hosts as x64 -- so derive the arch from
+# CMAKE_SYSTEM_PROCESSOR instead. See issue #2583.
+if(APPLE)
+  check_c_compiler_flag("-arch arm64" arm64Supported)
+else()
+  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|arm64")
+    set(arm64Supported TRUE)
+  else()
+    set(arm64Supported FALSE)
+  endif()
+endif()
 
 # TODO: if static runtime link is supported, then ABI tag postfix must include s
 # see:
@@ -32,7 +42,7 @@ else()
     set(DEBUG_POSTFIX "-d")
   endif()
   set(boost_LIB_PREFIX ${CMAKE_STATIC_LIBRARY_PREFIX})
-  if(${arm64Supported})
+  if(arm64Supported)
     set(PLATFORM_SUFFIX "a64")
   else()
     set(PLATFORM_SUFFIX "x64")

